src/sys/resource.rs

```diff
@@ -0,0 +1,230 @@
+/// println!("current hard_limit: {:?}", hard_limit);
+/// ```
+///
+/// # References
+///
+/// [getrlimit(2)](https://linux.die.net/man/2/getrlimit)
+///
+/// [`Resource`]: enum.Resource.html
+pub fn getrlimit(resource: Resource) -> Result<(Option<rlim_t>, Option<rlim_t>)> {
+    let mut old_rlim = rlimit {
+        rlim_cur: 0,
+        rlim_max: 0,
+    };
+
+    cfg_if! {
+        if #[cfg(all(target_os = "linux", target_env = "gnu"))]{
+            let res = unsafe { libc::getrlimit(resource as __rlimit_resource_t, &mut old_rlim) };
+        }else if #[cfg(any(
+            target_os = "freebsd",
+            target_os = "openbsd",
+            target_os = "netbsd",
+            target_os = "macos",
+            target_os = "ios",
+            target_os = "android",
+            target_os = "dragonfly",
+            target_os = "bitrig",
+            target_os = "linux", // target_env != "gnu"
+        ))]{
+            let res = unsafe { libc::getrlimit(resource as c_int, &mut old_rlim) };
+        }
+    }
+
+    Errno::result(res).map(|_| {
+        (
+            Some(old_rlim.rlim_cur).filter(|x| *x != RLIM_INFINITY),
+            Some(old_rlim.rlim_max).filter(|x| *x != RLIM_INFINITY),
+        )
+    })
+}
+
+/// Set the current processes resource limits
+///
+/// A value of `None` indicates that there's no limit.
+///
+/// # Parameters
+///
+/// * `resource`: The [`Resource`] that we want to set the limits of.
+/// * `soft_limit`: The value that the kernel enforces for the corresponding
+///   resource. Note: `None` input will be replaced by constant `RLIM_INFINITY`.
+/// * `hard_limit`: The ceiling for the soft limit. Must be lower or equal to
+///   the current hard limit for non-root users. Note: `None` input will be
+///   replaced by constant `RLIM_INFINITY`.
+///
+/// # Examples
+///
+/// ```no_run
+/// # use nix::sys::resource::{setrlimit, Resource};
+///
+/// let soft_limit = Some(1024);
+/// let hard_limit = None;
+/// setrlimit(Resource::RLIMIT_NOFILE, soft_limit, hard_limit).unwrap();
+/// ```
+///
+/// # References
+///
+/// [setrlimit(2)](https://linux.die.net/man/2/setrlimit)
+///
+/// [`Resource`]: enum.Resource.html
+pub fn setrlimit(
+    resource: Resource,
+    soft_limit: Option<rlim_t>,
+    hard_limit: Option<rlim_t>,
+) -> Result<()> {
+    let new_rlim = rlimit {
+        rlim_cur: soft_limit.unwrap_or(RLIM_INFINITY),
+        rlim_max: hard_limit.unwrap_or(RLIM_INFINITY),
+    };
+    cfg_if! {
+        if #[cfg(all(target_os = "linux", target_env = "gnu"))]{
+            // the below implementation is mimicing the similar implementation in golang
+            // https://go-review.googlesource.com/c/sys/+/230478/2/unix/syscall_linux_arm64.go#176
+            // seems for some of the architectures, we prefer to use prlimit instead of {g,s}etrlimit
+            let res = unsafe { libc::prlimit(0, resource as __rlimit_resource_t, &new_rlim as *const _, std::ptr::null_mut()) };
```

Comment:
I think this is just GoLang being silly.  Go religiously avoids using libc at all, prefering to make syscalls directly.  Rust doesn't do that.  You should just use setrlimit.

Reply:
The reason why I change to this it because in some of the architecture, using `setrlimit` in `libc` fails for unit test of set `RLIMIT_STACK` (test `test_resource_limits_stack`). To make all the architecture working, I did some research and found this implementation in golang. The failed architecture was `aarch64` and `arm` (see the [failure build](https://travis-ci.com/github/LMJW/nix/builds/187996674)). I am not quite sure why `setrlimit` is not working for some architectures. But using `prlimit` seems working for all the architectures. That's why I choose this implementation.

Reply:
I'm not satisfied.  Nix's goal is to make the system's C library Rusty.  Cross-platform portability and platform-specific workarounds are beyond our remit.  We've gotten in trouble in the past by trying to add this kind of compatibility code.  Please just use setrlimit alone.